packages/txservice/src/monitor/monitor.ts

```diff
@@ -0,0 +1,242 @@
+        throw result.error;
+      }
+      this.logger.warn(
+        {
+          method,
+          transaction: transaction,
+          error: result.error,
+        },
+        "Estimate gas failed due to an unexpected error.",
+      );
+      throw result.error;
+    } else {
+      gasLimit = result.value;
+    }
+
+    // Get gas price and create tracker instance.
+    result = await this.provider.getGasPrice();
+    if (result.isErr()) {
+      throw result.error;
+    }
+    const gas = new Gas(result.value, gasLimit);
+    return gas;
+  }
+
+  /**
+   * Get the current nonce value. Should ONLY ever be called within a serialized
+   * queue.
+   *
+   * @remarks
+   * Caller should still be prepared to get the incorrect nonce back. For instance,
+   * if the provider that just handled our sent tx has suddenly gone offline, this
+   * method may give the wrong nonce. This can be solved by making additional calls to
+   * submit the tx.
+   *
+   * @returns A number value for the current nonce.
+   */
+  private async getNonce(): Promise<number> {
+    // Update nonce value to greatest of all nonce values retrieved.
+    const buffer = this.buffer.getLastNonce() ?? -1;
+    const result = await this.provider.getTransactionCount();
+    if (result.isErr()) {
+      throw result.error;
+    }
+    // If transaction buffer above returns null, that indicates the buffer is empty; meaning
+    // we haven't sent any transactions yet, and we'll use this value here.
+    const pending = result.value;
+    // Increment if we got back a value from buffer as we want the value of last tx's nonce + 1.
+    return Math.max(buffer + 1, pending);
+  }
+
+  private async backfillLoop() {
+    // TODO: Make sure this loop is throw-proof
+    while (true) {
+      delay(1000);
+      const pending = this.buffer.pending;
+      // Lazy solution: we only care about a potential hold-up if it could hold anything up.
+      if (pending.length < 2) {
+        continue;
+      }
+      const result = await this.provider.getTransactionCount();
+      if (result.isErr()) {
+        // TODO: log
+      }
+      const currentNonce = result.isOk() ? result.value : -1;
+      const tx: Transaction | undefined = this.buffer.get(currentNonce);
+      if (tx == null) {
+        // BC it could just be the latest nonce and we're currently not doing anything
+        // TODO: backfill
+        this.backfill(currentNonce, undefined, "NOT_FOUND");
+      } else {
+        if (tx.didFinish || tx.isBackfill) {
+          // IF the transaction did finish already, or this is already being backfilled (from a previous iteration
+          // here), we can ignore this one.
+          // TODO: Do we actually want to proceed even if a tx IS a backfill? What would that accomplish? Because if
+          // even the backfill isn't working, then how would ANOTHER backfill solve anything? In fact, maybe we want to
+          // shut things down if even a backfill tx isn't going through?
```

Comment:
100% we should shut things down if the backfill tx isn't going through. This means we should make sure that the backfill is both awaited and submitted at a high gas price

Reply:
Currently we are submitting at max and we are awaiting

I think rn the question is how do we shut things down? Do we signal to parent provider class that this isn't working, and we should go down for maintenence?

It may be worth implementing a "non-functioning" hibernation mode for provider. Something that just polls every 10 minutes to see if provider is working again. Idk, future stuff